### ISSUE TITLE
wasm: metrics never expire and accumulate

### ISSUE THREAD

**Issue by kyessenov** [envoyproxy/envoy#14070]
Wasm filter implements the stats by allocating Envoy stats in the context and never de-allocating them. Since ABI encourages dynamic values, the stats accumulate and eventually leak (also, potentially exposing a security leak if stat values are untrusted).

**Comment by mathetake** [comment]
how about automatically dropping metrics which haven't been seen for, say, 10 minutes in Envoy host side? That is implementable and does not require the new ABI (which seems not coming soon). The only concern I can think of is we change the assumption on Envoy host implementation that "metric ids returned by define_metric ABI call are always valid (i.e. calling increment/recod_metric never returns NotFound for these metric IDs)", but I think it is acceptable and easy fix in user side.

wdyt? @bianpengyuan @kyessenov @PiotrSikora @mandarjog 

ref: https://github.com/istio/istio/issues/21377

**Comment by bianpengyuan** [comment]
> how about automatically dropping metrics which haven't been seen for,

Want to understand this better, by dropping metrics, you mean drop the whole metric definition, or just some stale time series? I think the common problem is that for a metric, some time series are stale and never cleaned up. The metric itself is still continuously reported.

**Comment by aweeks** [comment]
> how about automatically dropping metrics which haven't been seen for,

As currently implemented, the number of metrics (in the sense of metric name, crossed with labels), increases without bound.  This manifests as an ever-increasing number of metrics visible via the `/stats/prometheus` endpoint ([link](https://istio.io/latest/docs/ops/integrations/prometheus/)).

Previous versions of istio (<=1.7) [supported](https://istio.io/v1.7/docs/reference/config/policy-and-telemetry/adapters/prometheus/) a `metricsExpirationPolicy` option to help address this.  

This is especially problematic when dealing with ephemeral workloads that are associated with unique labels (for example pod name, etc. in a K8s environment), as once these workloads are terminated, these metrics remain forever even though they will never be updated again.

**Comment by mathetake** [comment]
@bianpengyuan I'm thinking about dropping just the stable series -  calling `decRefCount()` on metric objects which are not updated for x minutes by checking periodically y secs/mins.

**Comment by bianpengyuan** [comment]
> I'm thinking about dropping just the stale series

Ok, that sounds good to me. We did the same for opencensus library to stackdriver metric expiry.

**Comment by mathetake** [comment]
great, I will work on a PoC and let's see how it looks like.

**Comment by kyessenov** [comment]
Since Wasm filter maintains its own stats root, I think the proposed functionality is generally useful in other extensions. The current mechanism of allow-listing stats is not very flexible, so a general "managed" stats root library that controls stat retention could be used throughout envoy for cases when stats are not directly owned by config (e.g. gRPC service method stats). One tricky case to consider is that we may also want to wait until stats are flushed before deleting metric objects.

**Comment by juwatanabe** [comment]
Any updates on the POC?

**Comment by kyessenov** [comment]
There has not been much activity so this issue is looking for someone willing to commit to fixing Envoy ABI.

**Comment by pawankkamboj** [comment]
any update?

**Comment by kyessenov** [comment]
Yes, I finally got around to rewriting away from Wasm https://github.com/istio/istio/pull/41441, circumventing the whole ABI question. I haven't done expiry, but it seems fairly straightforward now.

**Comment by LugiaChang** [comment]
@kyessenov  Hello, I use wasm filter by EnvoyFilter. But when I keep deleting the EnvoyFilter or pod (
The pod ip will be sent to envoy through the plugin configuration for routing decisions), there seems to be a memory leak in envoy, the memory never goes down. Will it be related to this issue?  If not then what could be the reason?
<img width="872" alt="image" src="https://user-images.githubusercontent.com/22168057/236156076-7dfca387-260a-4cd0-9ccb-d204cad59c32.png">
[profile.pb.gz](https://github.com/envoyproxy/envoy/files/11395538/profile.pb.gz)

**Comment by shubham1172** [comment]
Facing the same issue. The WASM metrics continue to stay and accumulate. On deleting the listener and recreating again, the metrics counter continue to get updated from the last seen value.

Initial count.
```
# TYPE plugins_wasm_hello_world_request_headers_added counter
plugins_wasm_hello_world_request_headers_added{} 3
```

Delete listener and recreate, make a single call to the endpoint.
```
# TYPE plugins_wasm_hello_world_request_headers_added counter
plugins_wasm_hello_world_request_headers_added{} 4
```

**Comment by wonderflow** [comment]
facing this issue for golang plugin as well: https://github.com/envoyproxy/envoy/issues/37808
